DirectProgramming/DPC++FPGA/Tutorials/Features/fp_pragmas/src/fp_pragmas.cpp

```diff
@@ -0,0 +1,259 @@
+//==============================================================
+// Copyright Intel Corporation
+//
+// SPDX-License-Identifier: MIT
+// =============================================================
+#include <CL/sycl.hpp>
+#include <CL/sycl/INTEL/fpga_extensions.hpp>
+#include <iomanip>
+#include <iostream>
+#include <vector>
+
+// dpc_common.hpp can be found in the dev-utilities include folder.
+// e.g., $ONEAPI_ROOT/dev-utilities//include/dpc_common.hpp
+#include "dpc_common.hpp"
+
+using namespace sycl;
+using namespace std;
+
+#pragma clang fp contract(off)
+#pragma clang fp reassociate(off)
+constexpr double EPSILON = 1e-6;
+
+// Forward declare the kernel name in the global scope.
+class ContractOffKernel;
+class ReassociateOffKernel;
+class ContractFastKernel;
+class ReassociateOnKernel;
+
+// Launch a kernel on the device specified by selector.
+// The kernel's functionality is designed to show the
+// baseline performance without the fp contract pragma.
+void disable_contract(const device_selector &selector, const double a,
+                      const double b, const double c, const double d,
+                      double &res) {
+  try {
+    queue q(selector, dpc_common::exception_handler,
+            property::queue::enable_profiling{});
+
+    buffer<double, 1> bufferA(&a, 1);
+    buffer<double, 1> bufferB(&b, 1);
+    buffer<double, 1> bufferC(&c, 1);
+    buffer<double, 1> bufferD(&d, 1);
+    buffer<double, 1> bufferE(&res, 1);
+
+    event e = q.submit([&](handler &h) {
+      accessor accessorA(bufferA, h, read_only);
+      accessor accessorB(bufferB, h, read_only);
+      accessor accessorC(bufferC, h, read_only);
+      accessor accessorD(bufferD, h, read_only);
+      accessor accessorRes(bufferE, h, write_only, noinit);
+
+      // FPGA-optimized kernel
+      // Using kernel_args_restrict tells the compiler that the input
+      // and output buffers won't alias.
```

Comment:
Please use this comment (taken from the fpga_compile tutorial) with appropriate modifications here and everywhere you use kernel_args_restrict:

        // The "kernel_args_restrict" tells the compiler that a, b, and r
        // do not alias. For a full explanation, see:
        //    DPC++FPGA/Tutorials/Features/kernel_args_restrict